### ISSUE TITLE
integration test for lexical queries

### ISSUE THREAD

**Issue by balmas** [alpheios-project/components#147]
We need an integration test to which we can supply a file containing a list of words and have the lexical query sequence executed for each word in the file, receiving in response the data we use to construct the popup in order to validate the results for different parsers and lexicons and identify gaps in either the business logic or the services.

The interface to the integration test could be a command-line tool or a HTML ui which can be loaded in the browser  (both options would be ideal)

The results from each step of the lexical query process should be represented in the output.

We should be able to identify in the results for each word

- the homonyms returned by the morphological parser
- the short definitions returned by the lexicon service for each homonym
- the full definitions returned by the lexicon service for each homonym

Misses at every step should be flagged for filtering of results

We should be able to configure as run time parameters:
- language
- lexicon to use for short definitions
- lexicon to use for full definitions

It would be ideal to package this as a service which requires no knowledge of how to build the code to run it.

**Comment by balmas** [comment]
@irina060981 here's an issue to cover the task we discussed yesterday. Let me know if it isn't clear. Thanks!

**Comment by irina060981** [comment]
I have created HTML version (it could be placed on server) - it uses File API and papaparse for handling files.
Here it is working version - http://alpheious-test3.irina-sklyarova.ru/

The workflow:
1) **Choose file** - choose json file with words
example:
[
	{
	"targetWord": "cepit",
        "languageCode": "lat",
        "lexiconShortOpts": { "codes": ["ls"] },
        "lexiconFullOpts": { "codes": [] }
	}
]
There are dictionaries codes that are used for shortOpts and fullOpts codes .
All codes with urls and names are uploaded from libraryConfig.json
If codes are empty, that all dictionaries will be used.

2) **Upload data** - parsed data for checking

3) **Check data** - starts collecting data from lexical quires and placed results to the screen

4) **Download Morph Data** - downloads csv file from morph-client

5) **Download Short Defs** - download csv file from lexical-client for short definitions (different dictionaries in separated columns)

5) **Download Full Defs** - download html files from lexical-client for full definitions (different dictionaries in separated files)

**Comment by irina060981** [comment]
Screenshots:

**1 step**
![image](https://user-images.githubusercontent.com/12377640/43681876-934fec90-98a4-11e8-8d2e-0330b0695834.png)

**2 step**
![image](https://user-images.githubusercontent.com/12377640/43681880-a74b1526-98a4-11e8-855f-c2ba1b43ecf5.png)

**3 step**
![image](https://user-images.githubusercontent.com/12377640/43681883-b712954c-98a4-11e8-95ec-c761605e4823.png)

**Result** (show link in full definitions show full text)
![image](https://user-images.githubusercontent.com/12377640/43681888-e3db73d2-98a4-11e8-80be-b88f30617daf.png)

**Comment by irina060981** [comment]
Downloaded files for 
[
	{
		"targetWord": "cepit",
        "languageCode": "lat",
        "lexiconShortOpts": { "codes": ["ls"] },
        "lexiconFullOpts": { "codes": ["ls"] }
	},
	{
		"targetWord": "μοῦσα",
        "languageCode": "grc",
        "lexiconShortOpts": { "codes": ["lsj", "aut"] },
        "lexiconFullOpts": { "codes": ["lsj", "aut"] }
	},
	{
		"targetWord": "mare",
        "languageCode": "lat",
        "lexiconShortOpts": { "codes": ["ls"] },
        "lexiconFullOpts": { "codes": ["ls"] }
	},
	{
		"targetWord": "συνδέει",
        "languageCode": "grc",
        "lexiconShortOpts": { "codes": ["lsj", "aut"] },
        "lexiconFullOpts": { "codes": ["lsj", "aut"] }
	},
	{
		"targetWord": "ινώδους",
        "languageCode": "grc",
        "lexiconShortOpts": { "codes": ["lsj", "aut"] },
        "lexiconFullOpts": { "codes": ["lsj", "aut"] }
	}
]

**Comment by irina060981** [comment]
Downloaded files in archive
[Downloads.zip](https://github.com/alpheios-project/components/files/2259967/Downloads.zip)

**Comment by irina060981** [comment]
Here it is a repo
https://github.com/alpheios-project/lexical-tests

**Comment by irina060981** [comment]
There is a problem with command-line tool:
For lexical queries I need to use - alpheios-morph-client and alpheios-lexicon-client, but they use window.fetch for getting data from API,
But There is no window in command-line variant here.

So for using it in command-line we should create a node.js version of it with using some node specific fetching.

**Comment by irina060981** [comment]
As we talked on a meeting today : 

 I have added Download Failed Words button
![image](https://user-imag
es.githubusercontent.com/12377640/43721079-34236c24-99d5-11e8-9e21-a33f81689593.png)

It creates a csv file with such columns:
MorphClient - shows no if there are no data from morph-client
ShortLexical and FullLexical  - show dictionaries that don't have data for the word (depending on dictionaries pointed in the json file)

If the word has morphData and lexicalData for all pointed dictionaries - it won't be placed in this file.

![image](https://user-images.githubusercontent.com/12377640/43721090-39c614ba-99d5-11e8-8e31-563758498b07.png)

**Comment by irina060981** [comment]
I have added the ability to get translations from lemma-client
![image](https://user-images.githubusercontent.com/12377640/43838464-bfb22d90-9b5e-11e8-9d3f-da6fe7ec9efe.png)

**Comment by irina060981** [comment]
I need to make a little fix in data-models for translations
https://github.com/alpheios-project/data-models/pull/41

**Comment by irina060981** [comment]
And Alpheios translation service should be published somewhere - and url should be updated in lemma-client repo

**Comment by balmas** [comment]
One small request: if the user tries to upload an invalid JSON file, could we catch that error and display it to the user? Right now it just doesn't do anything (the error can be found in the debugging console). Thanks!

I will work on getting both the translation service and this app deployed this week.  We may still need a command line version to test large files of words. I'm currently testing it out by running it for a partner who wanted these results for a list of 3K plus words and am not sure if I'm going to run out of memory in the browser...

**Comment by irina060981** [comment]
I have added several simple checks with error messages:

1) **check for JSON.parse**
![image](https://user-images.githubusercontent.com/12377640/43840186-6586c01a-9b63-11e8-8d10-8a15d401ba03.png)

2) **check that data inside the file is an array**
![image](https://user-images.githubusercontent.com/12377640/43840252-92691f10-9b63-11e8-805c-73b981807129.png)

3) **check that every word blok has targetWord and languageCode properties**
![image](https://user-images.githubusercontent.com/12377640/43840288-b15f5038-9b63-11e8-80b4-2cc2fd230cd1.png)

![image](https://user-images.githubusercontent.com/12377640/43840326-c23d1b74-9b63-11e8-8e13-cf4d74000c59.png)

If lexiconShortOpts or lexiconFullOpts are absent - than all codes will be checked for the language of the word from libraryConfig.json

**Comment by irina060981** [comment]
I have added a list of translation's languages to libraryConfig.json - it could be increased or reduced. I have put the current list from lemma-client.

**Comment by balmas** [comment]
a few additional comments and questions

- I think I'd prefer the names of the download files to exclude spaces. 
- the full definitions download file is html rather than csv. Am I correct that this is intentional since the data is html?

**Comment by irina060981** [comment]
1) I have removed spaces from file names (updated gitrepo and a version on my server)

2) yes, as we receive data in a html from lexical-client  for full definitions (with a lot of tags) it will be hardly readable in other file formats; so I created it as html files with a simple table of contents and links to the start of it;

**Comment by balmas** [comment]
(also, fyi, I've deployed this now at http://morph.alpheios.net/lexical-tests/demo/)

**Comment by balmas** [comment]
Ok, so a list of the remaining changes that I think are needed for this:

- [ ] can we have the download files be comma or tab separated rather than ; ? (This is just for ease of import into spreadsheet -- the ; isn't standard)
- [ ] ability to download failures of each type of data separately from all other. E.g. we have failedMorph now. We'd like to able to get failedShortDefs, failedFullDefs, failedTranslations. 
- [ ] ability to skip short or full definition lookups or both
- [ ] command line interface 

I will enter separate issues into morph-client, lexicon-client and lemma-client for the wrapping of the window fetch request to enable the command line use

**Comment by irina060981** [comment]
- [ ] delimiter - I have changed the delimiter to tab (comma is not a good choice, as there are commas in definitions and other properties. Also I have added configJson and tabDelimiter as a embedded parameters, so you coul change them manually in index.html.

**Comment by irina060981** [comment]
- [ ] **skip short/full defs** I have added two variants:

  1) you could skip definition for some words by setting "lexiconShortOpts": false or "lexiconFullOpts": false
     so there are 3 variants for lexiconShortOpts  and lexiconFullOpts : 
     - "lexicon...Opts": false - skip check
     - "lexicon...Opts": {} or  "lexicon...Opts": { codes: [] }  - all dictionaries
     - "lexicon...Opts": { codes: ['code'] }  - check dictionaries from codes array
  2) you could globally skip shortDefs or fullDefs using checkboxes
![image](https://user-images.githubusercontent.com/12377640/43957461-3caf99c6-9ceb-11e8-959f-32df5c715e1f.png)

Also I have added disabling checkboxes for defs and translations if they are globally skipped.

**Comment by balmas** [comment]
fyi the translation service is now running at https://ats.alpheios.net

**Comment by irina060981** [comment]
For command-line version I have created:
  1) https://github.com/alpheios-project/lexical-tests-node
     in readme you could see the simple description how to work with it

  2) node build for data-models (used in lexical-tests-node)
      https://github.com/alpheios-project/data-models/tree/for-cmdtools

  3) node build for morph-client (used in lexical-tests-node)
      https://github.com/alpheios-project/morph-client/tree/for-cmdtools

  4) node build for lexical-client (used in lexical-tests-node)
      https://github.com/alpheios-project/lexicon-client/tree/for-cmdtools

  5) node build for lemma-client (used in lexical-tests-node)
      https://github.com/alpheios-project/lemma-client/tree/for-cmdtools

**Comment by irina060981** [comment]
simple use:

node cmdtool
* with default values *

node cmdtool params.json data/data3.json
* with current files *

**Comment by balmas** [comment]
This looks great! A few additional requests:

- [ ] it would be great if we could use webpack to produce a prebuilt script so that the only thing the user has to do is install and run
- [ ] related to the above request, for the node build of the clients, can we make a new preset for this in the node-build repository, and a new output file for it, so that, e.g we can do something like `node build-dev-node` which would execute, e.g. `standard --fix && node --experimental-modules ./node_modules/alpheios-node-build/dist/build.mjs all development lib-node` and would output, e.g. alpheios-morph-client.node.js and then we can use aliases in the webpack config to pull in the alpheios-morph-client-node.js file at build time.
- [ ] can we add a configuration option to the cmdtool.js to all the user to specify a maximum number of concurrent requests, and then fork multiple processes? This will be very helpful to use in load testing the services and will make it possible to run the tests more quickly.

**Comment by irina060981** [comment]
[ ] **webpack**
 - I have rebuild the application (change a little structure) for using it with webpack (backend config variant) correctly. 
-  I have updated https://github.com/alpheios-project/node-build/tree/for-cmdtools - to use preset node-app.
- I have added **npm run build** script

[ ] **execute with installed node.js"
- I have installed https://www.npmjs.com/package/pkg for creating self-executed packages for Linux, MacOS and Windows
- I have added **npm run package**

In both variants they could be used with inline arguments or without (will be got default values)

**Comment by irina060981** [comment]
I will work on creating and uses special builts with target: node for clients, data-models and use it in master branch (tomorrow)
I will work on creating concurrency async/await requests (using Promise.all) with max_query param ( I think I will place in params.json) (tomorrow)

**Comment by irina060981** [comment]
- [x]  **rebuild with node-lib preset**

1) I have created a new preset - node-lib in https://github.com/alpheios-project/node-build/pull/6

2) I have made rebuild for data-models - https://github.com/alpheios-project/data-models/pull/44

3) for morph-client - https://github.com/alpheios-project/morph-client/pull/21

4) for lexicon-client - https://github.com/alpheios-project/lexicon-client/pull/14

5) for lemma-client - https://github.com/alpheios-project/lemma-client/pull/5

6) I have updated lexical-tests and lexical-tests-node

I have tested rebuilt packages with lexical-tests, lexical-tests-node and embedded-lib - everything works

**Comment by irina060981** [comment]
- [x] Rebuiling command line version of lexical-tests to use concurency for requests:
  https://github.com/alpheios-project/lexical-tests-node

- change order of command-line arguments (data, params, config)
- refactor code for concurrency using requests
- added queue_max parameter to data.json (optional, default = 2)
- added concurrency requests (by user - similiar as for lexical-tests)

**Comment by balmas** [comment]
@irina060981 this is looking really great. 1 question and 1 request:

Question: I am not sure I understand your comment about the concurrency requests for the definitions. What were you trying to run in parallel that you were not able to? Do you mean that for each word in the data file, you wanted to be able to run the definition requests in parallel if there were multiple dictionaries to check?

Request: Can you please add documentation of all 3 configuration files to the readme of both lexical-tests and lexical-tests-node? The params file is documented in the readme of lexical-tests-node, already, missing the other 2 there, and all 3 in lexical-tests readme.

Thank you!

**Comment by irina060981** [comment]
@balmas:
**Definition request - (lexical-client)**
yes - I tried to make two variants parallel by user (done) and parallel by client-lib (not done).

I expect to do the second variant this way - 

1) get all words from morph-client in parallel by queue-max parameter (it is easy to do as it has one window-fetch for one word, so parallel library could easy to queue it)

2) get all words from lexical-client for short definitions (the same for full definitions) - I couldn't do that, because - request  using lexical client are made through two steps:
    a) `Lexicons.fetchShortDefs(lexeme.lemma, curOpts)` - it uses window.fetch in combination with SetTimeout - so it may be starts in parallel (I tried it) but I don't see it in network tab this way - it seems to me that first requests are recollected in 
`let adapters = Lexicons._filterAdapters(lemma, requestOptions)`
and then inits using additional function
`adapter[lookupFunction](lemma)`

So I think that parallel library doesn't catch correctly the start and the end of the request to init the next request at a proper moment (but it is only my thoughts I don't know how to check it).
  
   b) `let definition = await definitionRequest.request`
   It is really hard to define when the previous step completely finished so it is hardly defined when to start the second.
   
3) As for translations adapter - it works well as morph-client

For investigating this I have added the same functionality to web version of lexical-tests and you could easily see in a network Tab of the browser (I used Chrome) how it works in parallel mode (using skipShortDef, skipFullDef and not checking langs for translation - you could see different steps. 
It workd by user - so I use such a data file
[data3.zip](https://github.com/alpheios-project/components/files/2312580/data3.zip)

**Comment by irina060981** [comment]
Readme files updated with instructions and json files descriptions:
https://github.com/alpheios-project/lexical-tests
https://github.com/alpheios-project/lexical-tests-node

**Comment by balmas** [comment]
Yay! I made a few minor changes to the README and updated the sample data file. I'm marking this one done :-)

As people use it we may get additional requests, but these will be entered as issues on the lexical-tests repository. 

Great job on this @irina060981 . It's a great tool to have.

I will also enter a separate issue on the lexicon client for us to consider refactoring to better enable parallel requests.
